firefox-ios/Storage/Rust/RustFirefoxSuggest.swift

```diff
@@ -18,13 +18,21 @@ public protocol RustFirefoxSuggestActor: Actor {
 
     /// Interrupts any ongoing queries for suggestions.
     nonisolated func interruptReader()
+
+    /// Interrupts all ongoing operations.
+    nonisolated func interruptEverything()
 }
 
 /// An actor that wraps the synchronous Rust `SuggestStore` binding to execute
-/// blocking operations on the default global concurrent executor.
+/// blocking operations on a dispatch queue.
 public actor RustFirefoxSuggest: RustFirefoxSuggestActor {
```

Comment:
@linabutler I don't have full context and only took a quick look so I might have missed something, but the `actor` + GCD queues seems a bit confusing. At first glance I can't see how the separate read/write queues have any effect vs. using a single queue. My understanding is that the individual ingest/query calls would block until they completed (via `continuation.resume()`), meaning that the actor effectively serializes the execution anyway and so there would be no difference between 2 queues or one.

I tried creating a simplified test of this scenario outside of Firefox and at least the initial results appear to corroborate that; and I got identical output when I switched it to using a single queue (instead of 2).

More generally though, this approach of wrapping GCD/DispatchLib feels a bit like an anti-pattern. I worry that we're misusing some of the newer Swift Async features (though I don't know this area of the codebase - or Swift Async - well enough to be certain of how it should ideally be addressed).

By the way these comments aren't meant as a criticism of this PR, I'm mostly concerned about the bigger picture of our Swift Async adoption across the whole codebase. This seems like an example of some of the challenges/issues we'll be encountering elsewhere also.

Reply:
Hi @mattreaganmozilla, thank you so much for looking at this; it's very helpful—I didn't take it as a criticism at all, and your comments are great! ❤️ I hadn't used Swift actors at all before `RustFirefoxSuggest`, but I'm very familiar with GCD, so I think you have way more context than I do about the former.

> At first glance I can't see how the separate read/write queues have any effect vs. using a single queue.

That makes sense, and re-reading your explanation, that behavior sounds super obvious now! If I were doing this with vanilla GCD, I would use two separate queues, and have completion handlers for the operations.

Under the hood, in the Rust layer, `SuggestStore.query()` and `SuggestStore.ingest()` use two different SQLite connections, so it's possible for readers and writers to work concurrently all the way down.

But I forgot that the actor imposes its own serialization—and that's actually working against us here, because we don't need that; there's no mutable state to protect from concurrent modifications here!

I wonder if the "right" fix is to actually turn `RustFirefoxSuggest` into a plain `class`, with `async` methods that call `withContinuation`.

> More generally though, this approach of wrapping GCD/DispatchLib feels a bit like an anti-pattern. I worry that we're misusing some of the newer Swift Async features (though I don't know this area of the codebase - or Swift Async - well enough to be certain of how it should ideally be addressed).

To be honest, I haven't found good documentation about how to wire up Swift actors to blocking APIs—I'm wondering if you've turned up any?

All the examples and discussion threads that I've found show using one of the `withContinuation` variants to wrap a completion handler-based API, but not a blocking one.

Further, I've seen a lot of mentions that "tasks must make forward progress", which is why I suspected that calling the blocking `SuggestStore` methods from inside an actor method was the wrong thing to do. The `MXCPUException`s in Sentry make me suspect that we're blocking the main thread in the background ingestion task.

Reply:
@mattreaganmozilla I pushed a follow-up commit that turns `RustFirefoxSuggest` into a class, and keeps the uses of `withThrowingContinuation` to expose an `await`-compatible API that runs the tasks on the GCD queues. I think this is closer to how we'd wrap a completion handler-based API?